server/src/main/java/org/elasticsearch/cluster/ClusterModule.java

```diff
@@ -444,6 +447,8 @@ protected void configure() {
         bind(ShardsAllocator.class).toInstance(shardsAllocator);
         bind(ShardRoutingRoleStrategy.class).toInstance(shardRoutingRoleStrategy);
         bind(AllocationStatsService.class).toInstance(allocationStatsService);
+        bind(TelemetryProvider.class).toInstance(telemetryProvider);
+        bind(MetadataRolloverService.class).asEagerSingleton();
```

Comment:
I think this should be fine, but I am still unsure how did it work so that we were having multiple `MetadataRolloverService`. was this intentional to have multiple instances of this service?

Reply:
I suspect the intention was to have one instance. The reason I think this was okay is because MetadataRolloverService is almost a pure function, so there was no harm in having multiple copies. I will be sure to get another look from someone more familiar with this code before merging.

Reply:
> but I am still unsure how did it work so that we were having multiple MetadataRolloverService

The service was stateless so it wouldn't have mattered. 

I'm not super sure about this change. @pgomulka is this the general pattern for adding metrics? i.e. the services need to become stateful and keep track of their local counters? 
The pattern in my mind is going through the `MeterRegistry` to say something like:
```
registry.registerLongCounter( "myCounter", ... );

...

registry.getLongCounter("myCounter").incrementBy(1L, ...);
```
Am I missing something? 

Also, another question for the general output here - gauges are meant to go up and down - are we expecting the auto sharding metrics to decrease ?

Reply:
> is this the general pattern for adding metrics? i.e. the services need to become stateful and keep track of their local counters?
The pattern in my mind is going through the MeterRegistry to say something like:
registry.registerLongCounter( "myCounter", ... );
...
registry.getLongCounter("myCounter").incrementBy(1L, ...);

yes currently you need to have a state per a service that is creating metrics. We will be working to allow for static metric registration - think of logger - but it is not ready yet.

Reply:
@andreidan We discussed the alternatives a bit here: https://elastic.slack.com/archives/C8UUBNASY/p1713366670384779 .

My main reason for wanting to do the registration here is that it feels weird to do the registration up in NodeConstruction, then reference the same metric name far away in this code. Though I do see that adding state to this class is not ideal.

Reply:
> Also, another question for the general output here - gauges are meant to go up and down - are we expecting the auto sharding metrics to decrease ?

My thought is that we are not expecting the metrics to decrease. I believe `LongCounter` is for metrics that only increase, whereas `LongGauge` can decrease. Do you see a situation where we'd want the metrics to decrease?

Reply:
@pgomulka 

> yes currently you need to have a state per a service that is creating metrics. We will be working to allow for static metric registration - think of logger - but it is not ready yet.

Sorry if I'm missing something super obvious - but why do we need to have a Counter _both_ in the service and in the metrics registry? I understand a service must register its desired counters (e.g. doing this bit `registry.registerLongCounter( "myCounter", ... );` ) 
But what I don't understand is why do we need to keep hold of the counter in the service itself (like it's proposed [here](https://github.com/elastic/elasticsearch/pull/107593/files#diff-48f3bd0984e592b24ee2bd3e8e32b6f496054bd83e11755e8de156a202c2cf80R119) ) as opposed fetching it from the `MeterRegistry` ? (e.g. `registry.getLongCounter("myCounter").incrementBy(1L, ...);` ) 

The `autoShardingMetricCounters` state is added to `MetadataRolloverService` solely so we can do `autoShardingMetricCounters.get("myCounter").increment()`

Reply:
We discussed a bit, and I had misunderstood. It seems that @andreidan is correct that the `autoShardingMetricCounters` map is not needed. We can just do the `registry.registerLongCounter( "myCounter", ... );` in the constructor, not save the counters in a map, then access the counter with `registry.getLongCounter("myCounter").incrementBy(1L, ...);` shortly before emitting the metric. 

I had been thinking the suggestion was to move `registry.registerLongCounter( "myCounter", ... );` out of the class and up near NodeConstruction, to avoid making `MetadataRolloverService` a singleton. But it was just to remove the redundant state in `autoShardingMetricCounters`

Reply:
> Sorry if I'm missing something super obvious - but why do we need to have a Counter both in the service and in the metrics registry? I understand a service must register its desired counters (e.g. doing this bit registry.registerLongCounter( "myCounter", ... ); )
But what I don't understand is why do we need to keep hold of the counter in the service itself (like it's proposed [here](https://github.com/elastic/elasticsearch/pull/107593/files#diff-48f3bd0984e592b24ee2bd3e8e32b6f496054bd83e11755e8de156a202c2cf80R119) ) as opposed fetching it from the MeterRegistry ? (e.g. registry.getLongCounter("myCounter").incrementBy(1L, ...); )
The autoShardingMetricCounters state is added to MetadataRolloverService solely so we can do autoShardingMetricCounters.get("myCounter").increment()

you can do this, I think the effect is the same (one map less..) You would have to keep the state of the registry (a field somehow). So this is what I meant by keeping a state.
I think it is up to you to decide if you prefer to keep the fields local to the class or use a `registry.get`. Think of logger usage where you would keep the instance local to the class. 
Some parts of our codebase are 'aggregating' metrics as records - see S3RepositoriesMetrics usages.

The current state PR still looks good to me.